src/backend.c

```diff
@@ -5211,52 +5253,77 @@ int run_opencl_kernel_memset (hashcat_ctx_t *hashcat_ctx, hc_device_param_t *dev
 {
   const u64 num16d = size / 16;
   const u64 num16m = size % 16;
+  u32 tmp[4];
```

Comment:
I think the level of this array is different than in the other functions for CUDA and HIP.

Reply:
Reason is that hc_clEnqueueNDRangeKernel and hc_clEnqueueWriteBuffer both are now non-blocking and data needs to available until hc_clEnqueueWriteBuffer finishes at hc_clFinish.

Reply:
I see but then there's still an inconsistency with CUDA and HIP. Should it not be the same?

Reply:
> I see but then there's still an inconsistency with CUDA and HIP. Should it not be the same?

At the moment CUDA (and HIP as it's copy of CUDA) implementation is using legacy mode. That is in-order fully synchronized where each OP is synchronous.  We need to replace cuMemcpy functions with cuMemcpyAsync versions, which take stream as parameter. Currently memory copying is synchronized with default stream, 0. 

OpenCL had everything ready so I started with it. In-order FIFO execution guarantees correctness and we need to synchronize only when we need results from previous calculations.

I use OpenCL style:
Currently: Enqueue->Flush->Finish->Enqueue->Flush->Finish->memcpy (blocking)->Enqueue->Flush->Finish
Future: Enqueue->Enqueue->memcpy (non-blocking)->Enqueue->(Flush if needed to workaround)->Finish

So host execution is overlapping with device execution with asynchronous OPs as much as possible. Only synchronizing when we need results.

Reply:
OK,  lets start with OpenCL changes and see if there's any reports from users. After a while we should backport the changes to CUDA and HIP, too. What you think?

Reply:
Already working on CUDA and HIP